inductor: always convert weight to channels_last for cpu conv

XiaobingSuper opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #106122
* __->__ #105517


For the CPU backend, we always use channels_last to get good performance by avoiding format reorder(block to plain or plain to black), and they also assume that the weight is channels_last when doing the weight packing, so there always convert weight format and doing layout optimization.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

XiaobingSuper committed:
inductor: always convert weight to channels_last for cpu conv

[ghstack-poisoned]

jgong5 commented on test/inductor/test_inductor_freezing.py:
```diff
@@ -333,7 +333,7 @@ class Model(nn.Module):
             def __init__(self):
                 super().__init__()
                 self.kv = torch.nn.Conv2d(
-                    256, 384, kernel_size=(1, 1), stride=(1, 1), bias=False
```
better to cover both `group=1` and `group>1` cases?

XiaobingSuper committed:
Update on "inductor: always convert weight to channels_last for cpu conv"



For the CPU backend, we always use channels_last to get good performance by avoiding format reorder(block to plain or plain to black), and they also assume that the weight is channels_last when doing the weight packing, so there always convert weight format and doing layout optimization.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

shunting314 commented on torch/_inductor/graph.py:
```diff
@@ -227,6 +224,21 @@ def decide_layout_opt(gm) -> bool:
+        ):
+            return True
+
+        if not config.layout_optimization:
```
Can we move this back to the beginning of this method? This can be used as a kill switch (it's on by default). It would be weird if one disable layout_optimization (for whatever testing reason) and found it's still enabled due to cpu tensors.

XiaobingSuper replied:
Yes, updated.

XiaobingSuper committed:
Update on "inductor: always convert weight to channels_last for cpu conv"



For the CPU backend, we always use channels_last to get good performance by avoiding format reorder(block to plain or plain to black), and they also assume that the weight is channels_last when doing the weight packing, so there always convert weight format and doing layout optimization.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

shunting314 commented:
BTW, make sure the CI is clean before merging.

XiaobingSuper committed:
Update on "inductor: always convert weight to channels_last for cpu conv"



For the CPU backend, we always use channels_last to get good performance by avoiding format reorder(block to plain or plain to black), and they also assume that the weight is channels_last when doing the weight packing, so there always convert weight format and doing layout optimization.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

XiaobingSuper commented:
@pytorchbot merge

XiaobingSuper committed:
Update on "inductor: always convert weight to channels_last for cpu conv"



For the CPU backend, we always use channels_last to get good performance by avoiding format reorder(block to plain or plain to black), and they also assume that the weight is channels_last when doing the weight packing, so there always convert weight format and doing layout optimization.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

XiaobingSuper commented:
@pytorchbot merge
